test/functional/plugin/lsp_spec.lua

```diff
@@ -493,6 +493,34 @@ describe('LSP', function()
       eq(0, result.num_attached_after)
     end)
 
+    it('should not re-attach buffer if it was deleted in on_init #28575', function()
+      clear()
+      exec_lua(create_server_definition)
+      exec_lua([[
+        local server = _create_server({
+          handlers = {
+            initialize = function(method, params, callback)
+              vim.schedule(function()
+                callback(nil, { capabilities = {} })
+              end)
+            end
+          }
+        })
+        local bufnr = vim.api.nvim_create_buf(false, true)
+        local on_init_called = false
+        local client_id = vim.lsp.start({
+          name = 'detach-dummy',
+          cmd = server.cmd,
+          on_init = function()
+            vim.api.nvim_buf_delete(bufnr, {})
+            on_init_called = true
+          end
+        })
+        vim.lsp.buf_attach_client(bufnr, client_id)
+        vim.fn.wait(1000, on_init_called)
```

Comment:
```suggestion
        local ok = vim.wait(1000, function() return on_init_called end)
        assert(ok, "on_init was not called")
```

Reply:
Hmm actually this won't be enough either, because `on_attach` will be called after `on_init` so we will have to wait a bit more. (btw, the previous implementation is not evaluating `on_init_called` properly) It should be something like: until `on_attach` is called.

Reply:
> Hmm actually this won't be enough either, because on_attach will be called after on_init so we will have to wait a bit more.

I think it's fine because `on_init` and `_on_attach` are both called synchronously and should both run in the same event loop iteration.